### ISSUE TITLE
Optimizing out zero-width chunks

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#4173]
One potentially nice optimization would be to drop zero-width chunks. This would eliminate dependencies on other parts of the graph that are not used. Thinking about this mainly in the context of arrays and dataframes.

**Comment by mrocklin** [comment]
Where would this optimization take place?  When we construct Array objects?  How would we change the graph around in these cases?  Immediately?  Would this require an array name change?

**Comment by jakirkham** [comment]
Yeah it's difficult (at least for me) to see this not happening during `Array` construction as we don't seem to track this information afterwards. In which case you are right it probably needs an array name change. Maybe with some cleverness this could be avoided.

Perhaps we could start tracking this information somehow with a reachable flag. In which case this would become a dead graph elimination (akin to dead code elimination performed by compilers).

Maybe there are better strategies still.

**Comment by mrocklin** [comment]
My current thought is that maybe we can include this in the `top` logic so that every `top` operation ends up cleanup up chunks if any are zeroed out?  I have no idea if this is easy to do without adding a large amount of complexity though.

**Comment by jakirkham** [comment]
Do you have pointers on where you see this living in `top`?

Also what are your thoughts on having this in `Array` construction (given we've both now discussed this option 😉)?

**Comment by mrocklin** [comment]
> Do you have pointers on where you see this living in top?

Not off the top of my head, no.  This would require diving in a bit into the blockwise code (top was renamed to blockwise).

> Also what are your thoughts on having this in Array construction

I don't have any thoughts on this at the moment

**Comment by jakirkham** [comment]
My guess is this would be pretty easy to add to `Array` and would immediately benefit much of the codebase. That said, if there are concerns with that approach, would be interested in discussing them.

**Comment by mrocklin** [comment]
You would have to modify the passed in dictionary, which at times may be
challenging.

**Comment by jakirkham** [comment]
So perhaps the first step is simply to define a function that applies this operation to the Dask graph?

**Comment by mrocklin** [comment]
I don't know.  I'm not sure what exactly will happen when people try to do this.  I'm just portending that manipulating graphs at this stage might prove challenging.  People have already done a bunch of logic to make a graph.  You might have to undo and redo that logic in removing keys.
